### ISSUE TITLE
Non-Copying ents.GetAll(), player.GetAll() etc

### ISSUE THREAD

**Issue by jackson-c** [Facepunch/garrysmod-requests#738]
These functions return a _copy_ the whole table in question. This is useful (and safer) for getting a copy of the table then adding/removing some entities from it before processing each member. However, this is a BIG overhead and most uses of these functions don't attempt to modify the returned table. I think a non-copying variation of these functions (perhaps read-only?) would allow significant improvements to the speed of many scripts.

**Comment by VigiFyre** [comment]
This whole idea of table copying for "safety" is nothing but harm.

**Comment by zerfgog** [comment]
@VigiFyre What? How else will you prevent modification of reference types?

**Comment by thegrb93** [comment]
You can return a table who's metatable's __index metamethod points to the ents.GetAll table and __newindex points to an empty function.

**Comment by zerfgog** [comment]
@thegrb93 
```lua
local real_table = {["data we don't want modified"] = "unmodified"}
local proxy_table = setmetatable({}, {__index = function(_, k) return real_table[k] end, __newindex = function() return false end})

local function ents_GetAll()
    return proxy_table
end
--------
local t = ents_GetAll()
t["something new"] = "something new"
print(t["something new"]) -- nil (as you say)
-- HOWEVER
t["data we don't want modified"] = "modified"
print(t["data we don't want modified"]) -- modified (sad trombone)
```

**Comment by zerfgog** [comment]
@thegrb93 fair enough, but that is less practical than copying a table

**Comment by thegrb93** [comment]
Ah, then make the __newindex function revert the table to its current behavior by copying the ents and removing the metatable.

**Comment by zerfgog** [comment]
> copying the ents and removing the metatable

:+1:

**Comment by thegrb93** [comment]
pairs and ipairs won't work on it though... :/

**Comment by Kefta** [comment]
__pairs __ipairs (although the ladder is deprecated)

**Comment by thegrb93** [comment]
Ah cool, well there you go @willox . solved.

**Comment by willox** [comment]
seriously?

**Comment by thegrb93** [comment]
What? It allows reading the entity table without copying and when then user needs to edit the table, it makes a copy automatically. That's a solution no?

**Comment by zerfgog** [comment]
Lua is beautiful

**Comment by VigiFyre** [comment]
@zerfgog
> What? How else will you prevent modification of reference types?
Very simple.
If it costs performance, you don't.

**Comment by Kefta** [comment]
You know what costs performance? Making an entire copy of the entire ents/player.GetAll table every time it's called.

**Comment by jackson-c** [comment]
@VigiFyre Agreed. It's not like we need to protect the game from coders. If someone wants to screw up gmod, they can do so in thousands of ways anyway. So copying a bunch of data for "security purposes" is just silly.

**Comment by thegrb93** [comment]
Yeah, but after such a change, it'll be almost impossible to pinpoint where some addon creator decided it was a good idea to use table.remove to filter out entities they didn't want from the ents list.

**Comment by jackson-c** [comment]
@thegrb93 That's why I propose making alternate non-copying versions of these functions, while leaving the originals in place.

**Comment by Kefta** [comment]
Yes, let's please make more aliases and duplicates. Ultimately, this debate is pointless. Willox will decide what to do; just leave it alone.

**Comment by zerfgog** [comment]
@Kefta You misunderstood @VigiFyre's latest response, he was stating that we shouldn't have copying functions due to the performance hit they cause.

@thegrb93 This point exactly. Especially with people already used to removing entries from returned tables, it will be hard to fix these bugs.

@Kefta We have weapons.Get and weapons.GetStored for this exact purpose. They are used for their own reasons.

**Comment by LeQuackers** [comment]
Call me crazy, but couldn't you make a C++ function that calls a Lua defined function on every entity, such as ents.ForEach( func ), or would this be more/just as expensive? You wouldn't have a table to work with, but most of the time ents.GetAll is used in some kind of loop anyway.

**Comment by uRandomAlex** [comment]
I have also thought about kind of "EntityThink" hook that will call every Think for every entity. That way you are iterating over the whole entities table only once.

**Comment by Kefta** [comment]
@LeQuackers You would have to cross from C to Lua every function call. Would be much slower.

**Comment by NeatNit** [comment]
How about returning an iterator instead of a table?

``` Lua
for i, ent in ents.Pairs() do
```

The name Pairs doesn't really mean anything in this context but it's guaranteed to click with Lua coders better than something more accurate like ents.GetIterator would.
